### PR TITLE
[FLINK-35415][base] Fix compatibility with Flink 1.19

### PR DESCRIPTION
**Pull request by yuxiqian**
This closes FLINK-35092, FLINK-35120, FLINK-35316, FLINK-35415.

* Adds E2e testcases on recent Flink versions including 1.19
* Fixes CDC compatibility issue with 1.19 (due to breaking changes in @Internal classes)

This is a temporary workaround to avoid breaking compatibility with Flink 1.19. Proposal about upgrading Flink base (FLINK-35385) could be implemented later.

### CODE REVIEW

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/test/java/org/apache/flink/cdc/connectors/doris/sink/DorisMetadataApplierITCase.java [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+import static org.apache.flink.cdc.connectors.doris.sink.DorisDataSinkOptions.BENODES;
+import static org.apache.flink.cdc.connectors.doris.sink.DorisDataSinkOptions.FENODES;
+import static org.apache.flink.cdc.connectors.doris.sink.DorisDataSinkOptions.PASSWORD;
+import static org.apache.flink.cdc.connectors.doris.sink.DorisDataSinkOptions.SINK_ENABLE_BATCH_MODE;
+import static org.apache.flink.cdc.connectors.doris.sink.DorisDataSinkOptions.SINK_ENABLE_DELETE;
+import static org.apache.flink.cdc.connectors.doris.sink.DorisDataSinkOptions.USERNAME;
+
+/** IT tests for {@link DorisMetadataApplier}. */
+@RunWith(Parameterized.class)
+public class DorisMetadataApplierITCase extends DorisSinkTestBase {
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/test/java/org/apache/flink/cdc/connectors/doris/sink/DorisMetadataApplierITCase.java:69]: It's better to separate it into two issues, Or more simply, split it into two commitments.

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java [hunk 2]
```diff
@@ -150,4 +150,37 @@ private String generateSinkName(SinkDef sinkDef) {
         return sinkDef.getName()
                 .orElse(String.format("Flink CDC Event Sink: %s", sinkDef.getType()));
     }
+
+    private static <CommT> SimpleVersionedSerializer<CommT> getCommittableSerializer(Object sink) {
+        // TwoPhaseCommittingSink has been deprecated, and its signature has changed
+        // during Flink 1.18 to 1.19. Remove this when Flink 1.18 is no longer supported.
+        try {
+            return (SimpleVersionedSerializer<CommT>)
+                    sink.getClass().getDeclaredMethod("getCommittableSerializer").invoke(sink);
```
Comment (lvyanquan) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java:159]: Considering that we have had similar operations before in DataSinkWriterOperator, this change is acceptable for me.

**Comment by yuxiqian** [comment]
Thanks for @lvyanquan's review, I've split out commits based on FLINK ticket.

**Comment by yuxiqian** [comment]
Pushed another commit to resolve CI issue, could @leonardBang please re-trigger the CI? Thanks!

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/test/java/org/apache/flink/cdc/connectors/doris/sink/DorisMetadataApplierITCase.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    }
+
+    private List<Event> generateAddColumnEvents(TableId tableId) {
+        Schema schema =
+                Schema.newBuilder()
+                        .column(new PhysicalColumn("id", DataTypes.INT().notNull(), null))
+                        .column(new PhysicalColumn("number", DataTypes.DOUBLE(), null))
+                        .column(new PhysicalColumn("name", DataTypes.VARCHAR(17), null))
+                        .primaryKey("id")
+                        .build();
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/test/java/org/apache/flink/cdc/connectors/doris/sink/DorisMetadataApplierITCase.java:126]: It's better to list all type that we support currently.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksMetadataApplierITCase.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    }
+
+    private List<Event> generateAddColumnEvents(TableId tableId) {
+        Schema schema =
+                Schema.newBuilder()
+                        .column(new PhysicalColumn("id", DataTypes.INT().notNull(), null))
+                        .column(new PhysicalColumn("number", DataTypes.DOUBLE(), null))
+                        .column(new PhysicalColumn("name", DataTypes.VARCHAR(17), null))
+                        .primaryKey("id")
+                        .build();
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/test/java/org/apache/flink/cdc/connectors/starrocks/sink/StarRocksMetadataApplierITCase.java:96]: List all supported  types will be better too.

**Review by lvyanquan — APPROVED** [review]
LGTM.

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java [hunk 5]
```diff
@@ -152,4 +152,37 @@ private String generateSinkName(SinkDef sinkDef) {
         return sinkDef.getName()
                 .orElse(String.format("Flink CDC Event Sink: %s", sinkDef.getType()));
     }
+
+    private static <CommT> SimpleVersionedSerializer<CommT> getCommittableSerializer(Object sink) {
+        // TwoPhaseCommittingSink has been deprecated, and its signature has changed
```
Comment (leonardBang) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java:157]: we can add a //FIXME note

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java (hunk 2 above)
Comment (leonardBang) [reply flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java:161]: +1,  distribute CDC with various Flink versions in the future is recommended just like other flink external connectors

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java [hunk 6]
```diff
@@ -152,4 +152,37 @@ private String generateSinkName(SinkDef sinkDef) {
...
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+            throw new RuntimeException("Failed to get CommittableSerializer", e);
+        }
+    }
+
+    private static <CommT>
+            OneInputStreamOperatorFactory<CommittableMessage<CommT>, CommittableMessage<CommT>>
+                    getCommitterOperatorFactory(
+                            Sink<Event> sink, boolean isBatchMode, boolean isCheckpointingEnabled) {
+        // OneInputStreamOperatorFactory is an @Internal class, and its signature has changed
```
Comment (leonardBang) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/DataSinkTranslator.java:171]: ditto

**Review by leonardBang — COMMENTED** [review]
Thanks @yuxiqian for the nice work, the change looks good to me. I left minor comments. It will better if you can organize the commits to make each main change more clear e.g.  bump flink version could be in one separate commit, fix sink compatibility could be in another commit, and add e2e case could be in another commit.

**Comment by yuxiqian** [comment]
Done, seems I messed up the commit history during last rebase... Fixed it.

**Review by leonardBang — APPROVED** [review]
Thanks @yuxiqian for the update, LGTM

**Pull request merged**